File: src/0501-0600/506 - Relative Ranks/relative_ranks.go
package main

import (
	"fmt"
	"sort"
)

func findRelativeRanks(score []int) []string {
	sorted_score := append([]int(nil), score...)
	sort.Slice(sorted_score, func(i, j int) bool {
		return sorted_score[i] > sorted_score[j]
	})

	rank := make(map[int]string)

	for i, v := range sorted_score {
		switch i {
		case 0:
			rank[v] = "Gold Medal"
		case 1:
			rank[v] = "Silver Medal"
		case 2:
			rank[v] = "Bronze Medal"
		default:
			rank[v] = fmt.Sprintf("%d", i+1)
		}
	}

	result := make([]string, len(score))
	for i, v := range score {
		result[i] = rank[v]
	}

	return result
}

File: src/0501-0600/506 - Relative Ranks/relative_ranks.rs
impl Solution {
    pub fn find_relative_ranks(score: Vec<i32>) -> Vec<String> {
        let mut score_sorted = score.clone();
        score_sorted.sort_unstable_by(|a, b| b.cmp(a));

        let mut map = HashMap::new();
        for (i, s) in score_sorted.iter().enumerate() {
            map.insert(s, i + 1);
        }

        score
            .iter()
            .map(|s| {
                let rank = map.get(s).unwrap();
                match rank {
                    1 => "Gold Medal".to_string(),
                    2 => "Silver Medal".to_string(),
                    3 => "Bronze Medal".to_string(),
                    _ => rank.to_string(),
                }
            })
            .collect()
    }
}
